Breaking opm-core dependency in opm-output; new summary implementation

jokva opened the pull request:
Hi,

Let me prefix this patch set which spans several repos by stating the goals:
- We want user-configurable output of summary files etc.
- It would be nice to be able to write eclipse-compatible files outside of the simulator
- We currently have an inverted dependency (output now depends on core, should be the other way around)
- The code itself was rather worn and could use some maintenance. Several features were unused.

With that out of the way, a short synopsis of the major changes:
- The old summary support has been ripped out and replaced with the new (not quite complete) implementation.
- The multiplexing output writer has been removed as it was unused. If there is a need for a similar feature I'll take a fresh approach to this.
- The `OutputWriter` interface has gotten rewritten and does no longer depend on `core` data types.
- A (simple) format for Cell data, called `Solution`, is the information exchange specification required for output operations. Clients are expected to properly populate it.
- I have introduced a temporary utility file in core which assists in the transition. The features of that file are largely naïve and intended as a starting point for internal-to-output data conversion.
- WellState has gotten a virtual method `report` that populates a `data::Wells` structure.
- Internal helpers have been rewritten
- VAG grid support has been removed (completely unused).
- writeVTKdata is **disabled**, other VTK output should work as before.
- EclipseWriter no longer inherits from OutputWriter.

The PR has a lot of code, and changes ripples through both core and simulators; however, I will make sure that with all three HEADs it will compile and tests will run. I'm trying to keep this set a refactoring first of all, so behavioural changes will come later.

I would appreciate if most discussion regarding approach, interfaces etc. happens in this PR, and only direct code comments in the other PRs.

Edit by @joakim-hove:

opm-common: https://github.com/OPM/opm-common/pull/121
opm-core: https://github.com/OPM/opm-core/pull/1016
opm-simulators: https://github.com/OPM/opm-simulators/pull/688

jokva committed:
Output format for cell data.

jokva committed:
Replace SimulationTimer with time_t in writeInit

Instead of passing a massive and complex instance around, reduce
dependencies by directly passing the interesting time values to
writeInit.

jokva committed:
Remove SimulationTimer from RFTWriter

jokva committed:
Remove old Summary support.

jokva committed:
writeTimeStep takes time parameters explicitly

Removes the dependency on SimulationTimerInterface, and instead expects
to get the raw values from caller.

jokva committed:
Output takes data::Solution over SimDataContainer

To break dependencies between modules, and to properly define the input
format for the output facilities, make writeTimeStep take data::Solution
over SimulationDataContainer.

jokva committed:
Store SimulationDataContainer for compability

In a transition period for non-eclipse output, store a pointer to the
underlying SimulationDataContainer. Should be deprecated by writing
non-eclipse output with the new input types.

jokva committed:
Replace phaseUsage with the raw ert mask

jokva committed:
Remove unused headers

jokva committed:
Remove unused member variables.

jokva committed:
Moved member init to initialisation list

jokva committed:
Use unit conversion table from parser/units

jokva committed:
xwel member in Wells; init_list constructors

Adding the xwel member to data::Wells to support restart file output,
defining an exchange format with that in mind. For convenience,
introduces some initializer_list constructors as the buildup otherwise
is tedious and clumsy. Does not change much w.r.t. external interface.

jokva committed:
Replace WellState with data::Wells

jokva committed:
Remove unused multiplexer/static create feature

jokva committed:
Rewrite support for reading restart files

The support for reading restart files has been written in order to break
the dependency on opm-core. init_from_restart_file now returns
opm-output defined data types that clients, typically simulators, can
use to populate its own internal structures at will.

jokva committed:
Remove outdated comments

jokva committed:
Summary takes secs_elapsed, not per-step duration

To stay consistent with the interface exposed by the other
eclipse-writer components, the summary facilities takes its time elapsed
since simulation start, not on a per-step basis.

jokva committed:
Enable Summary support in EclipseWriter

jokva committed:
data::Wells::emplace

jokva committed:
More intuitive order of rates in data::Wells

jokva committed:
Refactor RTFWriter

Remove the opm-core dependency and re-implement the RFT writer. The
approach has been changed in the sense that we now store and keep alive
a file instance for the RFT file, instead of re-opening the file at
every time step.

jokva committed:
Store schedule ref for cleaner access.

jokva committed:
Reorder fields in EclipseWriter

jokva committed:
EclipseWriter header cleanup

Remove the unnecessary or unused includes. Consequence is well enum
conversions being no longer public static. To not break the tests the
functions have been implemented in writenumwells.

jokva committed:
Replace core/grid arg with dimensions in WriteECL

The grid structure was only used to read static size-related properties,
meaning the instance itself was rather uninteresting. Replace the grid
argument with raw ints.

jokva committed:
Simplify writing procedure

jokva committed:
Remove opm-core from writeECLData

jokva committed:
String fixup: core -> output

jokva committed:
Remove ParameterGroup include from CornerpointChopper

jokva committed:
Removed deprecated VAG support.

jokva committed:
Replace DataMap with std::map in writeVTK

jokva committed:
Disable UnstructuredGrid VTK output.

UnstructuredGrid is deeply tied to core, and is disabled for now.
writeVtk should be rewritten to use EclipseGrid as input or something
similar, but support is dropped in its current state.

jokva committed:
Remove core/WellType from EclipseWriter

jokva committed:
Replace OPM_THROW with throw.

jokva committed:
Remove opm-core dependency in RFT test; rename

Rewrites the RFT test to be independent of opm-core data types and
functions, and renames it to test_RFT since the class now represents the
full file, not some open-closing handle.

jokva committed:
Remove opm-core dependency in EclipseWriter test

jokva committed:
Remove opm-core dependency in Restart test; rename

Removes the opm-core dependency in the restart test, and renames the
file to be more in style with the new test set.

jokva committed:
Remove opm-core dependency in writenumwells test

jokva committed:
Remove details namespace; introduce anon namespace

There is no reason for these symbols to be externally visible, so
they're now instead contained in an anonymous namespace.

jokva committed:
Rewrite internal Keyword helper class

jokva committed:
Rewrite internal FileName helper

jokva committed:
Rewrite internal Restart helper

atgeirr commented:
I have not yet had the time to look at this carefully, and will not be able to until Friday. How urgently do you need feedback?

jokva commented:
Depends. A lot of things won't work properly yet, but concurrent output is broken in master anyway (and this approach would likely solve some of the fundamental problems).

There's still a lot of work to do and so far they only milestone is breaking up the dependency. Summary output isn't correct (although it should be an easy fix), and several other aspects **will** change.

jokva committed:
Rewrite internal Restart-Solution helper

jokva committed:
Rewrite internal Init helper

jokva committed:
Implement EclipseWriter as private; hide RFT

Hides all the EclipseWriter details and helper classes using pimpl,
which now enforces the non-copyability of EclipseWriter (which relies on
open and stateful file handles).

jokva committed:
Use throw over OPM_THROW

jokva committed:
Remove boost::to_upper_copy

jokva committed:
Replace boost::posix_time with time_t

The boost object was only used to calculate elapsed posix time anyway,
which is required (and already managed) by the callers. This makes the
interface slightly less complicated and removes an otherwise pointless
dependency.

jokva committed:
Remove the OutputWriter-Eclipse inheritance

This is-a relationship is never used, and the vtable is never leveraged.
The augmented output writers already use uniquely named output writers
anyway, so there is no reason to go via the hurdle of inheritance.

This gives a lot of flexibility when it comes to interface, since
Eclipse output can make Eclipse-specific assumptions.

jokva commented:
This builds now.

![364](https://cloud.githubusercontent.com/assets/8667957/15392281/5c2843f8-1dc6-11e6-8e7b-0fdf039007cf.png)

Locally I have modified opm-common to change the dependency order (i.e. core now depends on output), but these changes and test obviously work even if it "depends" on common. Nothing from common is now used in opm-output.

This does **not** mean that all PRs are mergable yet, but now opm-output can be inspected and reviewed more or less on its own.

jokva committed:
Remove unused argument from writeInit

jokva committed:
Read sim start time from EclipseState

Reading the start time from EclipseState means it's no longer required
as a parameter to every method, and that the current time can be
calculated as start time + time elapsed.

jokva commented:
I've changed the PRs in both core and simulators to be à jour with the recent development, and now the simulators and all works. I'd like someone to give travis a hint and build this with OPM/opm-common#121, ~~OPM/opm-parser#814~~, ~~OPM/opm-output#31~~, ~~OPM/opm-output#32~~, OPM/opm-core#1016 and OPM/opm-simulators#688

Oh, and review very welcome.

jokva commented:
I'd also like to add that the Summary output still isn't complete, but I hope that a switch and some thorough testing  here can uncover the bugs.

jokva committed:
Travis: don't build material and core

jokva committed:
Replace raw conversion_table with UnitSystem

The to_si/from_si functions were moved into UnitSystem which now manages
table lookup for clients, rather than having to store the raw
conversion tables.

joakim-hove commented on opm/output/Cells.hpp:
```diff
@@ -0,0 +1,84 @@
+        inline const std::vector< double >& operator[]( key ) const;
+        inline std::vector< double >& operator[]( key );
+
+        void insert( key, std::vector< double > );
```
Shouldn't this be: `void insert( key, std::vector< double > &&);`

joakim-hove replied:
For sanity checks it would make sense to check the size of the input vector; the expected number of cells could be a member variable specified in the constructor.

jokva replied:
Doesn't have to be, and doesn't mandate it. The operation itself will nearly always be a move anyway (especially if I get my shit together and make the definition inline which it should've been in the first place).

edit: but half the time this is the same vector that the simulator keeps around, so that just moves the copy burden over on them (which isn't great). An overload if anything.

joakim-hove commented on opm/output/Cells.hpp:
```diff
@@ -0,0 +1,84 @@
+         * * all units are SI
+         * * cells are active indexed
+         */
+        std::map< key, std::vector< double > > data;
```
I guess this should be private? At least when `has()` and `operator[]` has been implemented.

jokva replied:
Maybe, I'm not sure yet. I initially just aliased `std::map` to Wells, but later needed more so I wrapped a class around it. Now it should possibly be private, but it's exposed in order to be data "dumping" format (so to speak). I figured the answer would be clearer once everything started calming down around it, and this is good feedback w.r.t. that.

joakim-hove commented on opm/output/OutputWriter.hpp:
```diff
@@ -74,44 +75,28 @@ class OutputWriter {
      * timer.currentStepNum () == 0)
      */
-    virtual void writeInit(const SimulatorTimerInterface &timer, const NNC& nnc = NNC()) = 0;
+    virtual void writeInit( const NNC& nnc = NNC() ) = 0;
```
Has been much fighting with this NNC - please avoid default variable for now.

jokva replied:
Inherited and not changed, but I can remove it for sure.

pgdr replied:
Is that necessary?  The fighting with NNC has not been caused by this.  I furthermore think that a default NNC here makes sense, since if we don't have NNC information to write to output, it shouldn't be up to the caller of this method to create an empty NNC object.

jokva replied:
Quick check in simulators shows that the NNC is always passed regardless, although the discussion probably should go on.

joakim-hove commented on opm/output/OutputWriter.hpp:
```diff
@@ -52,10 +53,10 @@ struct PhaseUsage;
  *
  *  // before the first timestep
- *  writer->writeInit (timer);
+ *  writer->writeInit( current_posix_time, time_since_epoch_at_start );
```
`time_since_epoch_at_start` ~ is `seconds_elapsed` a better name.

jokva replied:
Old patch. The entire thing is gone.

jokva committed:
NNC variable not defaulted in writeInit

joakim-hove commented on opm/output/OutputWriter.hpp:
```diff
@@ -74,44 +75,28 @@ class OutputWriter {
+                                double seconds_elapsed,
+                                data::Solution reservoirState,
+                                data::Wells,
+                                bool  isSubstep) = 0;
```
We are semantically not really interested if this is a substep or not; the interesing question is:
1. Should I write summary data?
2. Should I write restart data?

The answer to the first question is "always" yes; although you can use the keyword `RPTONLY`[1] to essentially configure a `false` on this. For the second question we have logic like:

``` C++
if (!isSubstep)
   printRestart();
```

But then I think it is better to name the variable something like `printRestart` - which is the real question; in exceptional cases that can also be set to true even for a substep - although I do not think our full stack supports that at the moment.

[1]. I do not think we support that - quite uncommon.

jokva replied:
I totally agree and it's a much better interface.

I'll make a note of it and add it to the list of interface improvements (the `isSubstep` logic is inherited from the old implementation).

joakim-hove commented on opm/output/Wells.hpp:
```diff
@@ -150,7 +176,43 @@ namespace Opm {
+                );
+    }
+
+    inline Well& Wells::operator[]( const std::string& k ) {
```
Do we want the autovivification properties of `operator[]`?

jokva replied:
Maybe. It exists for clients so they have an easier time building in the case of many wells. I have been playing with the idea of forbidding to modify an established `Wells` though, i.e. all wells (and by extension connections) must be passed at construction time.

joakim-hove commented on opm/output/Wells.hpp:
```diff
@@ -88,7 +89,32 @@ namespace Opm {
+
+        std::map< std::string, Well > wells;
+        std::vector< double > bhp;
+        std::vector< double > temperature;
```
The temperature variable is orders of magnitude less relevant than the others. In most(?) of the cases it will only have a fixed dummy value - if not downright removed it should at least be removed to the end of the argument list.

jokva replied:
Maybe. This is slightly a preservation of the old approach, but more importantly the `bhp`, `temperature` and `wellrates` must all have the same size, where `perf` don't.

I do agree the constructor needs some work though.

joakim-hove replied:
> but more importantly the bhp, temperature and wellrates must all have the same size

I mess up this every second time; but I would have thought: `wellrates.size() == bhp.size() * numphases`?

jokva replied:
That's true, an oversight on my part. Which is why I put it after the `temperature` so the vectors are grouped by expected size.

joakim-hove commented on opm/output/eclipse/EclipseReader.cpp:
```diff
@@ -17,230 +17,163 @@
-                std::string error_str = "Restart file is missing SGAS data!\n";
-                throw std::runtime_error(error_str);
-            }
+        struct keyword {
```
Hopefully you could benefit from using: https://github.com/Ensembles/ert/blob/master/devel/libeclxx/include/ert/ecl/EclKW.hpp

jokva replied:
Yes, but it requires patching of EclKW. It's coming.

joakim-hove commented on opm/output/eclipse/EclipseReader.cpp:
```diff
@@ -17,230 +17,163 @@
-                std::string error_str = "Restart file is missing SWAT data!\n";
-                throw std::runtime_error(error_str);
-            }
+        for( const auto* key : { "PRESSURE", "TEMP", "SWAT", "SGAS" } ) {
```
This must be configurable; see previous comment about "TEMP". For the phases present the `phaseUsage.phase_used[BlackoilPhases::Aqua]` or similar should be consulted.

General comment:
1. The model always(?) has a water phase - i.e. `SWAT`.
2. A two-phase model will _only_ have water - the saturation of the _other_ phase, be that oil or gas, is implicitly inferred from `S_{other} = 1 - S_{wat}`.
3. For three phase models we enumarate water and gas saturations; if needed the oil saturation is `S_{oil} = 1 - S_{gas} - S_{wat}`.

jokva replied:
`PhaseUsage` is unavailable, and this is currently just a port of old behaviour. I've added it on the TODO list.

joakim-hove replied:
> PhaseUsage is unavailable, and this is currently just a port of old behaviour. I've added it on the TODO list.

OK; the essential information in the `phaseUsage` variable is available in `EclipseState` as well.

jokva replied:
That's true, and it's where I get the phase information from now.

joakim-hove commented on opm/output/eclipse/EclipseReader.cpp:
```diff
@@ -17,230 +17,163 @@
+        std::transform( sol[ ds::TEMP ].begin(), sol[ ds::TEMP ].end(),
+                        sol[ ds::TEMP ].begin(), apply_temperature );
+
+        /* optional keywords */
```
The `RS`and `RV`keywords are not really optional - if the model is specified with the `DISGAS` and `VAPOIL` keywords these keywords _should_ be present in the restart file.

joakim-hove commented on opm/output/eclipse/EclipseReader.cpp:
```diff
@@ -17,230 +17,163 @@
+        const double* xwel_data = ecl_kw_get_double_ptr(xwel);
+        const double* xwel_end  = xwel_data + ecl_kw_get_size( xwel );
+
+        const double* bhp_begin = xwel_data;
```
Holy diver - I thought pointer arithmetic was frowned upon?

jokva replied:
Is it? They're iterators in this context.

joakim-hove replied:
> They're iterators in this context

They are used as iterators - but surely pointer arithmetic is used; as long as it is tested it is fine by my though.

jokva replied:
In C++, pointers are a special case of iterators, implemented as a `ptrsize` integer (or whatever that constant is called, I never remember).

I've made a note of it though, and I'll consider changing it into something less surprising.

joakim-hove commented on opm/output/eclipse/EclipseReader.cpp:
```diff
@@ -17,230 +17,163 @@
+    int restart_step                     = initConfig->getRestartStep();
+    const std::string& restart_file_root = initConfig->getRestartRootName();
+    bool output                          = false;
+    const std::string& filename          = ioConfig->getRestartFileName(
```
The `getRestartFileName()` has value return (should probably a have a better name ....); so I think the filename should be a value as well.

jokva replied:
It does! I'll update the patch.

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -104,171 +77,98 @@ void restrictAndReorderToActiveCells(std::vector<double> &data,
 
-    void set(const std::string name, const std::vector<const char *>& data)
+    Keyword( const std::string& name, const std::vector< const char* >& data ) :
+        handle( ecl_kw_alloc( name.c_str(), data.size(), ECL_CHAR_TYPE ) )
```
This `handle` variable

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -104,171 +77,98 @@ void restrictAndReorderToActiveCells(std::vector<double> &data,
-    }
-
-    ecl_kw_type *ertHandle_;
+    ERT::ert_unique_ptr< ecl_kw_type, ecl_kw_free > handle;
```
Again - here I hope you can use: https://github.com/Ensembles/ert/blob/master/devel/libeclxx/src/EclKW.cpp - if not we should improve the ert class.

If it does not make sense to use the ert class please find a better name than `handle`[1] - handle to what??

[1]: I know the old used that name - and it always annoyed me.

jokva replied:
That's a nice addition, I'll look into it now.

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -104,171 +77,98 @@ void restrictAndReorderToActiveCells(std::vector<double> &data,
-    ~FileName()
-    { std::free(ertHandle_); }
+             bool formatted ) :
+        handle( ecl_util_alloc_filename(
```
And now `handle` is suddenly a `char *`\- what a mess this was.

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -296,277 +196,151 @@ class Restart : private boost::noncopyable
-    char *restartFileName_;
-    ecl_rst_file_type *restartFileHandle_;
+    FileName filename;
+    ERT::ert_unique_ptr< ecl_rst_file_type, ecl_rst_file_close > handle;
```
Again: `handle`

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -104,171 +77,98 @@ void restrictAndReorderToActiveCells(std::vector<double> &data,
 
 private:
-    char *ertHandle_;
+    using fd = std::unique_ptr< char, decltype( std::free )* >;
```
Better naming required - `fd == filedescriptor`; and again `handle`??

jokva replied:
`file descriptor` as is common in linux programs. Anywho, it was mostly to make the variable declaration neater, and no biggie.

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -296,277 +196,151 @@ class Restart : private boost::noncopyable
+                ioConfig.getUNIFOUT() ? ECL_UNIFIED_RESTART_FILE : ECL_RESTART_FILE,
+                writeStepIdx,
+                ioConfig.getFMTOUT() ),
+        handle(
```
Here `handle`seems to be a `ecl_file_type *` instance?

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -296,277 +196,151 @@ class Restart : private boost::noncopyable
+               ECL_EGRID_FILE,
+               writeStepIdx,
+               ioConfig.getFMTOUT() ),
+        init( fortio_open_writer(
```
Here you can look at: https://github.com/Ensembles/ert/blob/master/devel/libeclxx/include/ert/ecl/FortIO.hpp#L35

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -296,277 +196,151 @@ class Restart : private boost::noncopyable
     {
-        auto dataField = eclipseState->get3DProperties().getDoubleGridProperty("PORO").getData();
-        restrictAndReorderToActiveCells(dataField, numCells, compressedToCartesianCellIdx);
+        auto dataField = es.get3DProperties().getDoubleGridProperty("PORO").getData();
```
We really should leave it as is for now - to get this merged; but quite soon I would like to changes this:

``` C++
const auto& dataField = es.get3DProperties().getDoubleGridProperty("PORO").getData();
auto outputField = restrictAndReorderToActiveCells( dataField, numCells, compressedToCartesianCellIdx );
```

The reseating taking place in `restrictAndReorderToActiveCells()` is less than intuitive.

jokva replied:
I think the goal should be to get rid of restrict-and-reorder altogether, but that's on the horizon, and as you rightly say, out of the scope of this change set.

andlaus replied:
note that outputing _anything_ directly from EclipseState's grid properties should be considered to be bad style because these properties may be inconsistent with what the simulator actually uses: in this case (PORO), the simulator actually uses the pore volume which may or may not be calculated using PORO; in any case, the porosity gets modified by the rock compressibility, NTG and MULTPV, so the values which are put into the output files might not be what the user expects...

jokva replied:
This is for the INIT file though in which case we're interested in writing whatever is in the Deck (no?), rather than what the simulator sets up internally later.

Correct me if I'm wrong in this, but I'd assume that meant outputting whatever is provided by the Deck (and so EclipseState).

joakim-hove replied:
There is certainly room for improvement in this code; the current ambition is "just" to get it to compile and work with the opm-core <-> opm-output depency changed (i.e. opm-output should _not_ depend on opm-core). When 
that is fixed we can continue with improvements, but your initial claim:

> note that outputing anything directly from EclipseState's grid properties should be considered to be bad style because 

Is just plain wrong. 

You are right that the simulator uses `PORV` - which is either initialized from cell volumes, `PORO`and `NTG` or entered explicitly in the grid. We should output `PORV` - from the `Eclipse3DProperties` array[1]; and if present in the grid we should output `PORO` and also `NTG`. But the fact the simulator uses `PORV` does not in any way warrant the claim that we should not output `PORO`.

[1]: Actually the `PORV` vector is somewhat special in the INIT file, that vector _should_ be written as a vector of `nx * ny * nz` elements - where zero values can be used to infer the active-inactive cells without using a grid. We do not have that special casing in place yet (I think) - but that should be trivial to add.

andlaus replied:
I thought that the intention of the ECL restart files is to output what the simulator uses and not repeat what the deck says. Given that this is ECL, I'm not too much surprised if that is not the case, though. (I'm just wondering what the value proposition of that data is then.)

> You are right that the simulator uses PORV - which is either initialized from cell volumes, POROand NTG or entered explicitly in the grid. We should output PORV - from the Eclipse3DProperties array[1]; and if present in the grid we should output PORO and also NTG. But the fact the simulator uses PORV does not in any way warrant the claim that we should not output PORO.

do you mean that an additional check for PORO is needed because it is not present in some decks? In that case, I agree, but what happens for "partially defaulted" PORO fields where the defaulted cells are specified via PORV?

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -580,880 +354,418 @@ class Init : private boost::noncopyable
-    fortio_type *ertHandle_;
-    FileName egridFileName_;
+    FileName egrid;
+    ERT::ert_unique_ptr< fortio_type, fortio_fclose > init;
```
The name `init` says nothing about this being a file-like object. I would maybe assume it did not even need to be member -  but that can be handled in the future; the same applies to the `FileName egrid`.

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -580,880 +354,418 @@ class Init : private boost::noncopyable
 
+const int inactive_index = -1;
 
+/// Convert OPM phase usage to ERT bitmask
```
The `ECL_WATER_PHASE` and so on values are really Eclipse values and not ERT values.

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -580,880 +354,418 @@ class Init : private boost::noncopyable
-        // report the new production total
-        return unit::convert::to(total_, targetRateToSiConversionFactor_);
-    }
+    for( size_t active_index = 0; active_index < num_cells; ++active_index )
```
We are maintaning varieties of this mapping again and again; for the future we should try to use _only_ the grid - all the required mappings are there.

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -580,880 +354,418 @@ class Init : private boost::noncopyable
-            OPM_THROW(std::logic_error,
-                      "Unexpected unit type " << unitType);
+            const auto global_index = grid.getGlobalIndex( i, j, k );
+            const int index = this->global_to_active[ global_index ];
```
``` C++
if (!grid.cellActive(i,j,k))
   continue;
```

joakim-hove replied:
If https://github.com/OPM/opm-parser/pull/817 is merged you could do:

``` C++
if (!grid.cellActive(i,j,k))
   continue;
{
     size_t index = grid.activeIndex(i,j,k);
     const double depth = ...
     const double press = !pressure.empty() ? pressure[ index ] : 0.0;
     .....
}
```

joakim-hove commented on opm/output/eclipse/EclipseWriter.cpp:
```diff
@@ -580,880 +354,418 @@ class Init : private boost::noncopyable
-  return ert_well_type;
 }
 
+class EclipseWriter::Impl {
```
Please - can we stop this pattern?!

jokva replied:
What, the private implementation? It saves us exposing all the internal RFT/Restart/Init etc. to users, which is nice.

If you're talking about the name itself I don't have that strong an opinion (since it's isolated anyway), but I figured it was better than my original idea of "E" or "EW".

jokva committed:
Keyword replaced with ERT::EclKW

Use library-provided ecl_kw abstraction rather than rolling our own RAII
wrapper.

joakim-hove commented on tests/test_Restart.cpp:
```diff
@@ -0,0 +1,300 @@
+}
+
+BOOST_AUTO_TEST_CASE(EclipseReadWriteWellStateData) {
+    test_work_area_type * test_area = test_work_area_alloc("test_Restart");
```
This is available as: https://github.com/Ensembles/ert/blob/master/devel/libert_utilxx/include/ert/util/TestArea.hpp

jokva committed:
Renaming: 'handle' names made more descriptive

jokva committed:
Replace ert_unique_ptr with FortIO

jokva committed:
Relace ad-hoc keyword-struct with EclKW_ref

jokva committed:
Use grid-provided index calculations.

jokva committed:
Use alias for UnitSystem::measure

jokva committed:
Use basename for filename, not title

jokva committed:
Correctly calculate totals

jokva committed:
Base Summary test durations on days

Basing the dest durations on days over seconds make them less noisy,
more readable and closer to actual idiomatic use. Removes some subtle
complexities from the tests.

jokva committed:
Don't crash when a Well is not in data::Wells

It's perfectly legitimate for clients to "default" a well result, e.g.
when it hasn't started up yet, in which case most output will default to
0. Avoid crashing when this happens, and instead calculate based on a
dummy well.

jokva commented:
![norne-snap-output](https://cloud.githubusercontent.com/assets/8667957/15850550/e7dc27c6-2c98-11e6-8a68-9e28057e4246.png)

edit: looks like there might be an off-by-one error in rates here, I'll look into this.

jokva committed:
Look up history rates at correct timestep

The input format and parser interacts non-intuitively, as the observed
rates for timestep N are written for timestep N-1.

jokva committed:
Only write positive values to Summary file

Since production is seen as negative injection, and production rates
often are zero, negative zeros would be returned. Likewise, occasionally
numerical noise gives slightly-below-zero values which are now rounded
to be plain zeros.

jokva commented:
This should be ready to merge now.

![norne-snap-output1](https://cloud.githubusercontent.com/assets/8667957/15958060/86d96562-2ef3-11e6-9946-90d39f9994fd.png)

bska commented:
This PR breaks the build of module opm-upscaling.  Given the magnitude of the change that is probably not entirely unexpected but we still need to fix it.  In particular, class [`Opm::SteadyStateUpscalerImplicit<>`](https://github.com/OPM/opm-upscaling/blob/master/opm/upscaling/SteadyStateUpscalerImplicit.hpp#L52) is not properly converted to account for changes in the [`writeECLData`](https://github.com/OPM/opm-output/blob/master/opm/output/eclipse/writeECLData.hpp#L32-L39) interface of commit 43ce7fc.

I do not know if there are other build issues in module opm-upscaling.

jokva commented:
I'll have a look, but it hasn't changed behaviour so it should just be an interface upgrade.

jokva commented:
@bska https://github.com/OPM/opm-upscaling/pull/194 Mind checking it out?
